## Debugging

<details>
<summary>Failed linking step of libiec61850.dylib</summary>

```shell
libiec61850> [100%] Linking C shared library libiec61850.dylib
libiec61850> cd /tmp/nix-build-libiec61850-1.6.0.drv-0/source/build/src && /nix/store/bhlxbydiiasxicqxyd3vzh502dx421ql-cmake-3.31.6/bin/cmake -E cmake_link_script CMakeFiles/iec61850-shared.dir/link.txt --verbose=1
libiec61850> /nix/store/xpja3ad45p55aryvgng6g9grp8yivvsd-clang-wrapper-19.1.7/bin/clang  -Wredundant-decls -Wundef -O3 -DNDEBUG -isysroot /nix/store/afinrggjfrk7q9pb7rv50l50cj0y0l63-apple-sdk-11.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 1.6.0 -o libiec61850.1.6.0.dylib -install_name @rpath/libiec61850.1.6.0.dylib "CMakeFiles/iec61850-shared.dir/common/string_map.c.o" "CMakeFiles/iec61850-shared.dir/common/map.c.o" "CMakeFiles/iec61850-shared.dir/common/linked_list.c.o" "CMakeFiles/iec61850-shared.dir/common/byte_buffer.c.o" "CMakeFiles/iec61850-shared.dir/common/string_utilities.c.o" "CMakeFiles/iec61850-shared.dir/common/buffer_chain.c.o" "CMakeFiles/iec61850-shared.dir/common/conversions.c.o" "CMakeFiles/iec61850-shared.dir/common/mem_alloc_linked_list.c.o" "CMakeFiles/iec61850-shared.dir/common/simple_allocator.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_server/iso_connection.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_server/iso_server.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_acse/acse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/common/mms_type_spec.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/common/mms_value.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/common/mms_common_msg.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_initiate.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_write.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_identify.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_status.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_named_variable_list.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_connection.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_files.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_get_namelist.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_get_var_access.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_common.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_read.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/client/mms_client_journals.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_read_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_file_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_association_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_identify_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_status_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_named_variable_list_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_value_cache.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_get_namelist_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_access_result.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_server.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_server_common.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_named_variable_list.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_domain.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_device.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_information_report.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_journal.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_journal_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_server_connection.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_write_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/server/mms_get_var_access_service.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_cotp/cotp.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_presentation/iso_presentation.c.o" "CMakeFiles/iec61850-shared.dir/mms/asn1/ber_decode.c.o" "CMakeFiles/iec61850-shared.dir/mms/asn1/asn1_ber_primitive_value.c.o" "CMakeFiles/iec61850-shared.dir/mms/asn1/ber_encoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/asn1/ber_integer.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_client/iso_client_connection.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_common/iso_connection_parameters.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_session/iso_session.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/client/client_control.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/client/client_report_control.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/client/client_goose_control.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/client/client_sv_control.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/client/client_report.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/client/ied_connection.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/common/iec61850_common.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/impl/ied_server.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/impl/ied_server_config.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/impl/client_connection.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/model/model.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/model/dynamic_model.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/model/cdc.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/model/config_file_parser.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/mms_mapping/control.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/mms_mapping/mms_mapping.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/mms_mapping/reporting.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/mms_mapping/mms_goose.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/mms_mapping/mms_sv.c.o" "CMakeFiles/iec61850-shared.dir/iec61850/server/mms_mapping/logging.c.o" "CMakeFiles/iec61850-shared.dir/logging/log_storage.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/DataAccessError.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/DeleteNamedVariableListRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/constr_SET_OF.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/MmsPdu.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GetNamedVariableListAttributesResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/BIT_STRING.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ber_tlv_tag.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/constr_SEQUENCE_OF.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/asn_SET_OF.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ReadResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/InformationReport.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConfirmedServiceRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/DeleteNamedVariableListResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/asn_SEQUENCE_OF.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/VariableAccessSpecification.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GetVariableAccessAttributesRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/xer_support.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ObjectName.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/NativeEnumerated.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/per_encoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/constr_SEQUENCE.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GetNameListResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/MMSString.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/InitiateErrorPdu.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/IndexRangeSeq.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConfirmedErrorPDU.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/UnconfirmedService.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/UTF8String.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ServiceError.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/TimeOfDay.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GetNameListRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/asn_codecs_prim.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Data.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ScatteredAccessDescription.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ReadRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/per_decoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Identifier.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ServiceSupportOptions.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Integer8.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConfirmedServiceResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ParameterSupportOptions.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Integer16.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ber_tlv_length.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/OCTET_STRING.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/DefineNamedVariableListRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/FloatingPoint.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/xer_encoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Unsigned8.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/BOOLEAN.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/INTEGER.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/UnconfirmedPDU.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/DataSequence.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/constraints.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/der_encoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/VisibleString.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/InitiateResponsePdu.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/StructComponent.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Address.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Unsigned16.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ber_decoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/per_support.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/WriteResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/InitRequestDetail.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/InitiateRequestPdu.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/DefineNamedVariableListResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/NULL.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ListOfVariableSeq.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/UtcTime.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConcludeResponsePDU.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/AccessResult.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Integer32.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GetNamedVariableListAttributesRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/VariableSpecification.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/Unsigned32.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/constr_CHOICE.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/AlternateAccess.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ObjectClass.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/InitResponseDetail.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConfirmedResponsePdu.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GetVariableAccessAttributesResponse.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/NativeInteger.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/xer_decoder.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/AlternateAccessSelection.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConfirmedRequestPdu.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/ConcludeRequestPDU.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/WriteRequest.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/RejectPDU.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/TypeSpecification.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/constr_TYPE.c.o" "CMakeFiles/iec61850-shared.dir/mms/iso_mms/asn1c/GeneralizedTime.c.o" "CMakeFiles/iec61850-shared.dir/goose/goose_subscriber.c.o" "CMakeFiles/iec61850-shared.dir/goose/goose_receiver.c.o" "CMakeFiles/iec61850-shared.dir/goose/goose_publisher.c.o" "CMakeFiles/iec61850-shared.dir/sampled_values/sv_subscriber.c.o" "CMakeFiles/iec61850-shared.dir/sampled_values/sv_publisher.c.o" "CMakeFiles/iec61850-shared.dir/r_session/r_session.c.o" "CMakeFiles/iec61850-shared.dir/r_session/r_session_crypto_mbedtls.c.o" "CMakeFiles/iec61850-shared.dir/sntp/sntp_client.c.o"  ../hal/libhal-shared.a
libiec61850> Undefined symbols for architecture arm64:
libiec61850>   "_mbedtls_ctr_drbg_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ctr_drbg_init", referenced from:
libiec61850>       _RSessionCrypto_createRandomData in r_session_crypto_mbedtls.c.o
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ctr_drbg_random", referenced from:
libiec61850>       _RSessionCrypto_createRandomData in r_session_crypto_mbedtls.c.o
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_setOwnKey in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_setOwnKeyFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ctr_drbg_seed", referenced from:
libiec61850>       _RSessionCrypto_createRandomData in r_session_crypto_mbedtls.c.o
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_entropy_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_entropy_func", referenced from:
libiec61850>       _RSessionCrypto_createRandomData in r_session_crypto_mbedtls.c.o
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_entropy_init", referenced from:
libiec61850>       _RSessionCrypto_createRandomData in r_session_crypto_mbedtls.c.o
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_gcm_auth_decrypt", referenced from:
libiec61850>       _RSessionCrypto_gcmAuthAndDecrypt in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_gcm_crypt_and_tag", referenced from:
libiec61850>       _RSessionCrypto_gcmEncryptAndTag in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_gcm_free", referenced from:
libiec61850>       _RSessionCrypto_gcmEncryptAndTag in r_session_crypto_mbedtls.c.o
libiec61850>       _RSessionCrypto_gcmAuthAndDecrypt in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_gcm_init", referenced from:
libiec61850>       _RSessionCrypto_gcmEncryptAndTag in r_session_crypto_mbedtls.c.o
libiec61850>       _RSessionCrypto_gcmAuthAndDecrypt in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_gcm_setkey", referenced from:
libiec61850>       _RSessionCrypto_gcmEncryptAndTag in r_session_crypto_mbedtls.c.o
libiec61850>       _RSessionCrypto_gcmAuthAndDecrypt in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_free", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_hmac_finish", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_hmac_starts", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_hmac_update", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_info_from_type", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_init", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_md_setup", referenced from:
libiec61850>       _RSessionCrypto_createHMAC in r_session_crypto_mbedtls.c.o
libiec61850>   "_mbedtls_pk_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_pk_init", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_pk_parse_key", referenced from:
libiec61850>       _TLSConfiguration_setOwnKey in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_pk_parse_keyfile", referenced from:
libiec61850>       _TLSConfiguration_setOwnKeyFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_psa_crypto_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_cache_free", referenced from:
libiec61850>       _TLSConfiguration_addCRL in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_addCRLFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_cache_get", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_cache_init", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_cache_set", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_cache_set_timeout", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_close_notify", referenced from:
libiec61850>       _TLSSocket_close in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_authmode", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_ca_chain", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSSocket_read in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSSocket_write in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_ciphersuites", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_endpoint", referenced from:
libiec61850>       _TLSConfiguration_setClientMode in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_max_version", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_min_version", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_own_cert", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_renegotiation", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_rng", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_session_cache", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_conf_verify", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_config_defaults", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_config_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_config_init", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_free", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSSocket_close in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_get_ciphersuite", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_get_session", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_get_verify_result", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSSocket_read in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_handshake", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_read", referenced from:
libiec61850>       _TLSSocket_read in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_renegotiate", referenced from:
libiec61850>       _TLSSocket_performHandshake in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_session_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_session_reset", referenced from:
libiec61850>       _TLSSocket_performHandshake in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSSocket_write in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_set_bio", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_set_session", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_setup", referenced from:
libiec61850>       _TLSSocket_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_ssl_write", referenced from:
libiec61850>       _TLSSocket_write in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crl_free", referenced from:
libiec61850>       _TLSConfiguration_resetCRL in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crl_init", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_resetCRL in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crl_parse", referenced from:
libiec61850>       _TLSConfiguration_addCRL in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crl_parse_file", referenced from:
libiec61850>       _TLSConfiguration_addCRLFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crt_free", referenced from:
libiec61850>       _TLSConfiguration_destroy in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crt_info", referenced from:
libiec61850>       _verifyCertificate in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crt_init", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crt_parse", referenced from:
libiec61850>       _TLSConfiguration_setOwnCertificate in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_addAllowedCertificate in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_addCACertificate in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_mbedtls_x509_crt_parse_file", referenced from:
libiec61850>       _TLSConfiguration_setOwnCertificateFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_addAllowedCertificateFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>       _TLSConfiguration_addCACertificateFromFile in libhal-shared.a(tls_mbedtls.c.o)
libiec61850>   "_psa_crypto_init", referenced from:
libiec61850>       _TLSConfiguration_create in libhal-shared.a(tls_mbedtls.c.o)
libiec61850> ld: symbol(s) not found for architecture arm64
libiec61850> clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
</details>

## Analysis

I believe the root cause is libiec61850's  broken `hal/CMakeLists.txt` which adds the mbedtls libraries via `target_link_libraries()` only for the `hal` but not the `hal-shared` library:

https://github.com/mz-automation/libiec61850/blob/3a1480e5948ca53b4dcc5fa15b05598fa3322bb9/hal/CMakeLists.txt#L204-L206

I have fixed this with a patch in nixpkgs here: https://github.com/stv0g/nixpkgs/commit/15457c6f4cf986d8dcaec386689fa19de280a204

I created an upstream PR here: https://github.com/mz-automation/libiec61850/pull/557

--- a/pkgs/by-name/li/libiec61850/package.nix
+++ b/pkgs/by-name/li/libiec61850/package.nix
@@ -1,9 +1,11 @@
 {
   cmake,
+  mbedtls,
   lib,
   stdenv,
   fetchFromGitHub,
   gitUpdater,
+  withTls ? !stdenv.hostPlatform.isDarwin,
 }:
 
 stdenv.mkDerivation (finalAttrs: {
@@ -19,7 +21,14 @@ stdenv.mkDerivation (finalAttrs: {
 
   separateDebugInfo = true;
 
+  cmakeFlags = lib.optionals withTls [
+    "-DCONFIG_USE_EXTERNAL_MBEDTLS_DYNLIB=ON"
+    "-DCONFIG_EXTERNAL_MBEDTLS_DYNLIB_PATH=${mbedtls}/lib"
+    "-DCONFIG_EXTERNAL_MBEDTLS_INCLUDE_PATH=${mbedtls}/include"
+  ];
+
   nativeBuildInputs = [ cmake ];
+  buildInputs = lib.optionals withTls [ mbedtls ];
 
   passthru.updateScript = gitUpdater { rev-prefix = "v"; };
 
@@ -27,7 +36,10 @@ stdenv.mkDerivation (finalAttrs: {
     description = "Open-source library for the IEC 61850 protocols";
     homepage = "https://libiec61850.com/";
     license = lib.licenses.gpl3Only;
-    maintainers = with lib.maintainers; [ stv0g pjungkamp ];
+    maintainers = with lib.maintainers; [
+      stv0g
+      pjungkamp
+    ];
     platforms = lib.platforms.unix;
   };
 })

